src/nbl/video/COpenGL_Swapchain.h

```diff
@@ -0,0 +1,202 @@
+#ifndef __NBL_C_OPENGL__SWAPCHAIN_H_INCLUDED__
+#define __NBL_C_OPENGL__SWAPCHAIN_H_INCLUDED__
+
+#include "nbl/video/ISwapchain.h"
+#include "nbl/video/COpenGLFunctionTable.h"
+#include "nbl/system/IThreadHandler.h"
+#include "nbl/video/COpenGLImage.h"
+#include "nbl/video/surface/ISurfaceGL.h"
+
+namespace nbl {
+namespace video
+{
+
+template <typename FunctionTableType_>
+class COpenGL_Swapchain final : public ISwapchain
+{
+    static inline constexpr uint32_t MaxImages = 4u;
+
+public:
+    using ImagesArrayType = ISwapchain::images_array_t;
+    using FunctionTableType = FunctionTableType_;
+
+    // should be called by GL/GLES backend's impl of vkQueuePresentKHR
+    void present(uint32_t _imgIx)
+    {
+        m_threadHandler.requestBlit(_imgIx);
+    }
+
+    static core::smart_refctd_ptr<COpenGL_Swapchain<FunctionTableType>> create(SCreationParams&& params, const egl::CEGL* _egl, ImagesArrayType&& images, COpenGLFeatureMap* _features, EGLContext _master, EGLConfig _config, EGLint _major, EGLint _minor)
+    {
+        if (!images || !images->size())
+            return nullptr;
+        if (images->size() < params.minImageCount)
+            return nullptr;
+        if (images->size() > MaxImages)
+            return nullptr;
+
+        auto extent = (*images)[0]->getCreationParameters().extent;
+        for (auto& img : (*images))
+        {
+            auto& ci = img->getCreationParameters();
+            if (ci.type != asset::IImage::ET_2D)
+                return nullptr;
+            if (ci.arrayLayers != params.arrayLayers)
+                return nullptr;
+            if (ci.mipLevels != 1u)
+                return nullptr;
+            if (ci.extent.width != extent.width)
+                return nullptr;
+            if (ci.extent.height != extent.height)
+                return nullptr;
+        }
+
+        return core::make_smart_refctd_ptr<COpenGL_Swapchain<FunctionTableType>>(std::move(params), _egl, std::move(images), _features, _master, _config, _major, _minor);
+    }
+
+    E_ACQUIRE_IMAGE_RESULT acquireNextImage(uint64_t timeout, IGPUSemaphore* semaphore, IGPUFence* fence, uint32_t* out_imgIx) override
+    {
+        // TODO impl
```

Comment:
you mean impl the semaphore and fence stuff?

Reply:
yep